File: super-egg-drop.cpp
//Time - O(KN)
//Space - O(KN)
class Solution {
public:
    int superEggDrop(int K, int N) {
        
        vector<vector<int>> dp(N+1, vector<int>(K+1,0));
        int attempts = 0;
        while(dp[attempts][K]<N){
            attempts++;
            for(int i = 1;i<=K;i++){
                dp[attempts][i] = 1 + dp[attempts-1][i-1] + dp[attempts-1][i];
            }
        }
        
        return attempts;
    }
};

File: super-egg-drop.js
// Time Complexity : O(K*N) but it has some early stopping power, and will never take the full N as a factor
// Space Complexity : O(K*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, this is a hard problem.  It was hard to wrap my head around the reverse process of
// mapping attempts and eggs to max floors

const superEggDrop = (K, N) => {
    const dp = new Array(N + 1);
    for (let i = 0; i < dp.length; i++) {
         dp[i] = new Array(K + 1).fill(0);
    }
    
    for (let attempt = 1; attempt <= N; attempt++) {
        for (let egg = 1; egg <= K; egg++) {
            dp[attempt][egg] = 1 + dp[attempt - 1][egg - 1] + dp[attempt - 1][egg];
            if (dp[attempt][egg] >= N) return attempt;
        }
    }
};
